File: GhostBot/functions/regen.py
from __future__ import annotations

import time
from typing import TYPE_CHECKING

from GhostBot import logger
from GhostBot.config import RegenConfig
from GhostBot.functions.runner import Locational

if TYPE_CHECKING:
    from GhostBot.bot_controller import ExtendedClient


class Regen(Locational):
    def __init__(self, client: ExtendedClient):
        super().__init__(client=client)

        self.config: RegenConfig = self._client.config.regen
        self._mana_threshold = float(self.config.mana_threshold or 0.75)
        self._hp_threshold = float(self.config.hp_threshold or 0.75)

    def _run(self) -> bool:
        """
        :return: True is we healed successfully, False if we were attacked, or in battle while healing
        """
        if self._mana_low() or self._hp_low():
            self._goto_start_location()

            if not self._client.in_battle:
                logger.info(f'{self._client.name}: low hp/mana, starting Regen')

                # mana/hp pots\
                if self.config.bindings.get('hp_pot') is not None:
                    if self._client.hp_percent < self._hp_threshold:
                        self._use_hp_pot()
                if self.config.bindings.get('mana_pot') is not None:
                    if self._client.mana_percent < self._mana_threshold:
                        self._use_mana_pot()

                hp = int(self._client.hp)
                while (self._client.hp < self._client.max_hp or self._client.mana < self._client.max_mana) and self._client.running:
                    logger.debug(f'{self._client.name}: healing')
                    time.sleep(2)
                    if self._client.in_battle or self._client.hp < hp:
                        logger.debug(f'{self._client.name}: Ouch, attacking')
                        return False
                    self._goto_spot_and_sit()
                    hp = int(self._client.hp)
                return True
        return False

    def _mana_low(self) -> int:
        return self._client.mana_percent < self._mana_threshold

    def _hp_low(self) -> int:
        return self._client.hp_percent < self._hp_threshold

    def _use_hp_pot(self) -> None:
        self._goto_spot_and_sit()
        self._client.press_key(self.config.bindings.get('hp_pot'))

    def _use_mana_pot(self) -> None:
        self._goto_spot_and_sit()
        self._client.press_key(self.config.bindings.get('mana_pot'))

    def _goto_spot_and_sit(self) -> None:
        self._goto_start_location()
        self._sit()

    def _sit(self):
        if not self._client.sitting:
            logger.debug(f'{self._client.name}: sitting')
            self._client.press_key(self.config.bindings.get('sit'))

File: kotlin-server/src/main/kotlin/com/ghostbot/functions/Regen.kt
package com.ghostbot.functions

import com.ghostbot.controller.BotClientWindow
import com.ghostbot.lib.seconds
import com.ghostbot.lib.sleepSeconds

/**
 * Port of GhostBot/functions/regen.py.
 */
class Regen(client: BotClientWindow, private val fairyActivated: Boolean = false) :
    Locational(client) {

    private val config = client.config?.regen ?: error("regen config required")
    private val manaThreshold: Double = (config.manaThreshold as? Double) ?: 0.75
    private val hpThreshold: Double = (config.hpThreshold as? Double) ?: 0.75

    override fun _run(): Boolean {
        // :return: true if we healed successfully, false if we were attacked
        // or in battle while healing
        if (manaLow() || hpLow()) {
            gotoStartLocation()

            val startWait = System.currentTimeMillis()
            if (client.inBattle) {
                var waitedOut = false
                while (client.inBattle && System.currentTimeMillis() - startWait < seconds(sec = 3) * 1000) {
                    sleepSeconds(0.5)
                    if (!client.inBattle) {
                        waitedOut = true
                        break
                    }
                }
                // Python while-else: loop ended by the 3s timeout (still in battle)
                if (!waitedOut && client.inBattle) return false
            }
            _logInfo("low hp/mana, starting Regen")

            val bindings = config.bindingsMap()
            if (bindings.isNotEmpty()) {
                // mana/hp pots
                useHpPot(bindings)
                useManaPot(bindings)
            }

            val currentHp = client.hp ?: return false
            var hp = currentHp
            val waitCondition: () -> Boolean = if (fairyActivated) {
                { (client.mana ?: 0) < (client.maxMana ?: 0) }
            } else {
                { (client.mana ?: 0) < (client.maxMana ?: 0) || (client.hp ?: 0) < (client.maxHp ?: 0) }
            }
            while (waitCondition() && client.running) {
                _logDebug("healing")
                sleepSeconds(2.0)
                if (client.inBattle || (client.hp ?: 0) < hp) {
                    _logDebug("Ouch, attacking")
                    return false
                }
                gotoSpotAndSit()
                hp = client.hp ?: return false
            }
            return true
        }
        return false
    }

    private fun manaLow(): Boolean = client.manaPercent() < manaThreshold

    private fun hpLow(): Boolean {
        if (fairyActivated) return false
        return client.hpPercent() < hpThreshold
    }

    private fun useHpPot(bindings: Map<String, Any?>) {
        if (client.hpPercent() < hpThreshold) {
            val key = bindings["hp_pot"]
            if (key != null) {
                gotoSpotAndSit()
                client.pressKey(key)
            }
        }
    }

    private fun useManaPot(bindings: Map<String, Any?>) {
        if (client.manaPercent() < manaThreshold) {
            val key = bindings["mana_pot"]
            if (key != null) {
                gotoSpotAndSit()
                client.pressKey(key)
            }
        }
    }

    private fun gotoSpotAndSit() {
        gotoStartLocation()
        sit()
    }

    private fun sit() {
        if (!client.sitting) {
            _logDebug("sitting")
            client.sit(config.bindingsMap()["sit"])
        }
    }
}
